libyul/optimiser/DataFlowAnalyzer.cpp

```diff
@@ -261,17 +255,17 @@ void DataFlowAnalyzer::handleAssignment(set<YulString> const& _variables, Expres
 	auto const& referencedVariables = movableChecker.referencedVariables();
 	for (auto const& name: _variables)
 	{
-		m_references.set(name, referencedVariables);
+		m_references[name] = referencedVariables;
 		if (!_isDeclaration)
 		{
 			// assignment to slot denoted by "name"
-			m_storage.eraseKey(name);
+			m_storage.erase(name);
 			// assignment to slot contents denoted by "name"
-			m_storage.eraseValue(name);
+			cxx20::erase_if(m_storage, [&](auto const& entry) { return entry.second == name; });
```

Comment:
I think it would be helpful to have a named utility function for this. It's just too easy to confuse things here.

Reply:
On the other hand, maybe it's better to go over `m_storage` just once and combine this with line 262?

Reply:
Sorry, but I'm not really sure what you mean :-). This line *is* calling a named utility function :-). You mean it's confusing that the "entry.second" is the value of the map, resp. the slot contents? To me it seems that extracting anything here would make this *harder* to read, not easier, but maybe I'm just not quite sure what you mean...

Reply:
> On the other hand, maybe it's better to go over `m_storage` just once and combine this with line 262?

It's not the same kind of going over the it - the first one can take advantage of it being a hash and removes keys faster, yielding a smaller map to run over in the second one quickly. I actually tried and to my own surprise it was noticably slower to combine them - if I recall correctly at least. I could try again and remeasure, though...

Reply:
On the other hand rather weird that the one entry more or less would actually make a difference, though... I'll recheck and see what happens...

Reply:
Sure, there are more comparisons to be made! What I meant was that it is not obvious at first glance what this line does, and a helper function called `eraseValues` would make that clearer.

Reply:
I'd actually argue the opposite that this line makes it perfectly obvious what it does, because everything is right there, whereas such a helper function would mean that I need to check what it actually does first...

Reply:
... with the difference that you already get a hint of what it is supposed to be doing. Using `.second` is always very confusing and I have to juggle with three things at a time (entry, name, second) to see what is going on.

Reply:
Also: This is a lambda function with capturing, so the first thing I have to do is check what the variables in there are actually referencing in the current scope. If it's a free function, this problem is gone, or at least the scope is much smaller.

Reply:
We could define something like
```
template<typename Callable>
decltype(auto) mapPair(auto&& callable) { return [&](auto&& pair) { return callable(pair.first, pair.second); }
```
And make it ``cxx20::erase_if(m_storage, mapPair([&name](YulString /* key */, YulString value) { return value == name; }));`` ;-)...

Reply:
That might solve the ``.second`` thing in more generality... and makes explicit that it doesn't really bind much...